ramips: Add support for Xiaomi MiWiFi 3A

minax007 opened the pull request:
The Xiaomi MiWiFi 3A wireless router has a similar system architecture as the Xiaomi Mi 4A (100M) router, which is already officially supported by OpenWrt.

Product website: https://www.mi.com/miwifi3a

Device specification
--------------------
SoC:      MT7628AN MIPS_24KEc @ 580 MHz 2.4G-bgn 2x2
WiFi:     MT7612EN 5G-an, ac 80 MHz 2T2R
Flash:    16 MB 
DRAM:     64 MB 
Switch:   MT7628AN (integrated in SoC)
Ethernet: 1 x 10 /100 Mbps
USB:      None
Antennas: 2 x 2,4 GHz and 2 x 5 GHz (all are external and non-detachable)
LEDs:     blue/red/amber
Buttons:  Reset
Serial:   115200,8n1

MAC addresses as verified by OEM firmware:
------------------------------------------
**use address source**
LAN *:DD factory 0x28
WAN *:DD factory 0x28
2g *:DE factory 0x4
5g *:DF factory 0x8004

OEM firmware uses VLAN's to create the network interface for WAN and LAN.

Bootloader info:
----------------
The stock bootloader uses a "Dual ROM Partition System".  OS1 is a deep copy of OS2.
The bootloader starts OS2 by default.
To force start OS1 it is needed to set "flag_try_sys2_failed=1".

How to install:
---------------
1- Use OpenWRTInvasion to gain Telnet, SSH and FTP access: https://github.com/acecilia/OpenWRTInvasion 
[IP: 192.168.31.1 | Username: root | Password: root | FTP-Port: 21] 
2- Connect to router using Telnet or SSH.
3- Backup all partitions. Use command "dd if=/dev/mtd0 of=/tmp/mtd0". Copy /tmp/mtd0 to computer using FTP. 
4- Copy openwrt-ramips-mt76x8-xiaomi_miwifi-3a-squashfs-sysupgrade.bin to /tmp in router using FTP. 
5- Enable UART access and change start image to OS1.

```
nvram set uart_en=1
nvram set flag_last_success=1
nvram set boot_wait=on
nvram set flag_try_sys2_failed=1
nvram commit
```

6- Erase OS1 & OS2 and install OpenWrt

```
mtd erase OS1
mtd erase OS2
mtd -r write /tmp/openwrt-ramips-mt76x8-xiaomi_miwifi-3a-squashfs-sysupgrade.bin OS1
```
Credits:
--------
This PR is based on the work of Zehao Zhang (Github: @ZZH-Finalize) that he had published in the PR: https://github.com/openwrt/openwrt/pull/15698

Signed-off-by: Olgun Demir <olgun.demir@mail.com.tr>

minax007 commented:
The PR works with the stock u-boot bootloader as demanded by the OpenWrt team.

I have compiled the firmware and tested it. It works without any issues since days. 

But in case someone breaks his device he can easily recover it by the stock u-boot bootloader. 

MiWifi-Repair-Tool does not work. 

The stock u-boot bootloader includes a recovery function. 

How to unbrick the Xiaomi Mi-3A:

1. Disconnect the power supply of the router
2. Push the reset button and hold it
3. Power up the router
4. Wait until the yellow LED starts flashing (takes approx. 8 seconds). Then release the reset button. 
5. Do the following setting for the LAN of your PC:

IP-address: 192.168.31.100
Subnet mask: 255.255.255.0
Standard gateway: 192.168.31.1
DNS server: 192.168.31.1

6. Connect the LAN cable to LAN port 2 of the router. 
7. Enter following address in your webbrowser
http://192.168.31.1

8. Upload the stock firmware
9. Wait until the router has rebooted and configure the router
10. Start the OpenWrt installation process as stated in the PR again

musashino205 commented on package/boot/uboot-envtools/files/ramips:
Please apply the change for your device to [package/boot/uboot-tools/uboot-envtools/files/ramips](https://github.com/openwrt/openwrt/blob/9ec997b85009e5357d5f21ad579578d068af8fb7/package/boot/uboot-tools/uboot-envtools/files/ramips) instead of this file.

minax007 replied:
Hi @musashino205 
Thanks for your response!
I will try to update the PR accordingly.

minax007 replied:
Same here is stated below regarding the old files that I have used. 
I will download the latest versions from the repository and update those accordingly. 
Thanks for pointing out to this!

minax007 replied:
I have updated the files.

musashino205 replied:
Please don't update this file and add the change to **[package/boot/uboot-tools/uboot-envtools/files/ramips](https://github.com/openwrt/openwrt/blob/9ec997b85009e5357d5f21ad579578d068af8fb7/package/boot/uboot-tools/uboot-envtools/files/ramips)** instead.
This file (package/boot/uboot-envtools/files/ramips) is not used now and should be deleted.

musashino205 commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -29,26 +29,6 @@ define Build/ravpower-wd009-factory
-  DEVICE_MODEL := WLR-1240
-endef
-TARGET_DEVICES += 7links_wlr-1240
-
```
Why did you delete or change definitions of many devices?

minax007 replied:
Sorry, that was not my intention. 
The code I took was from the former PR that is almost one year old as stated above. 
And I assume that many devices were included since then. 
I will download the latest versions of the files and do the adaptions. 
Sorry once again ... I didn't think about this when using the old files :-)

minax007 replied:
I have updated the files.

musashino205 commented:
Please use the cloned local git repository for modifications of codes instead of GitHub WebUI.

musashino205 commented:
> I was only using the following cloned repository and not the WebUI:
> https://github.com/minax007/openwrt
>
> Is there anything that I need to do different?

Ah OK, I misunderstood something...

---

Please squash the all commits to single commit.

minax007 commented:
Hi @musashino205 
I have squashed all the commits to a single commit.

Neustradamus commented:
To follow this nice PR :)

musashino205 commented:
> Please don't update this file and add the change to [package/boot/uboot-tools/uboot-envtools/files/ramips](https://github.com/openwrt/openwrt/blob/9ec997b85009e5357d5f21ad579578d068af8fb7/package/boot/uboot-tools/uboot-envtools/files/ramips) instead.
> This file (package/boot/uboot-envtools/files/ramips) is not used now and should be deleted.

minax007 commented:
> > Please don't update this file and add the change to [package/boot/uboot-tools/uboot-envtools/files/ramips](https://github.com/openwrt/openwrt/blob/9ec997b85009e5357d5f21ad579578d068af8fb7/package/boot/uboot-tools/uboot-envtools/files/ramips) instead.
> > This file (package/boot/uboot-envtools/files/ramips) is not used now and should be deleted.

Hi @musashino205 
thanks for your patience with me and referring again to your first comment of yesterday. 
Now I have understood that the structure has changed and have put the file to the correct path.

minax007 commented:
@musashino205 
At the bottom of this page I see the message 
"This branch is out-of-date with the base branch. 
Changes can be cleanly merged."

Do I need to take care of this?

When I click on update branch it creates a further commit. As I am not the originator it is not signed by me - and the automated pull request checks show a failure. 
Due to this I have rolled back to the previous state where you can only see my commit. 
Is this ok?

Shine- commented:
This:
> Please use the cloned local git repository for modifications of codes instead of GitHub WebUI.

When rebasing, rebase your local clone and then force-push.

minax007 commented:
> This:
> 
> > Please use the cloned local git repository for modifications of codes instead of GitHub WebUI.
> 
> When rebasing, rebase your local clone and then force-push.

Hi @Shine- 

Thanks for your answer and support! 
Since having understood the comment I am commiting via force push. 
Is it necessary to rebase - or is it ok to leave the pull request as it is?

Shine- commented:
As long as it can be cleanly merged (ie. no "fuzz" during `git am`) and no updates to files that are affected by other patches, there shouldn't be a need to rebase.
But you better rely on frequent contributors for details, not just me :-)

musashino205 commented on package/boot/uboot-tools/uboot-envtools/files/ramips:
```diff
@@ -64,6 +64,7 @@ mediatek,linkit-smart-7688|\
 samknows,whitebox-v8|\
 xiaomi,mi-router-4c|\
 xiaomi,miwifi-nano|\
+xiaomi,miwifi-3a|\
```
Please sort alphabetically. (`3a` shoud be placed before `nano`)

```
$ echo -e "nano\n3a"
nano
3a
$ echo -e "nano\n3a" | sort
3a
nano
```

minax007 replied:
I have made the change and have uploaded the updated file.

musashino205 commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -1200,6 +1200,15 @@ define Device/xiaomi_miwifi-nano
 endef
 TARGET_DEVICES += xiaomi_miwifi-nano
 
+define Device/xiaomi_miwifi-3a
+  IMAGE_SIZE := 16064k
+  DEVICE_VENDOR := Xiaomi
+  DEVICE_MODEL := MiWiFi 3A
+  DEVICE_PACKAGES := kmod-mt76x2
+  SUPPORTED_DEVICES += miwifi-3a
+endef
+TARGET_DEVICES += xiaomi_miwifi-3a
+
```
Please move this before `xiaomi_miwifi-nano`.

minax007 replied:
I have made the change and have uploaded the updated file.

minax007 replied:
@musashino205 
I have checked also all the other files and have shifted the "miwifi 3a" related lines not only before "miwifi nano" but also before "miwifi 3c".

minax007 commented:
Hi @musashino205 
I did not get any feedback from you for a week. 
Please let me know if there is anything else I need to do. 
Thanks!

musashino205 commented on target/linux/ramips/dts/mt7628an_xiaomi_miwifi-3a.dts:
```diff
@@ -0,0 +1,181 @@
+&wmac {
+	status = "okay";
+
+	mediatek,mtd-eeprom = <&factory 0x0>;
```
Please convert to the NVMEM binding from `mediatek,mtd-eeprom`.

minax007 replied:
Thanks for your fast response!
I have adapted the file accordingly. 
Please check if anything else needs to be changed. 
Thanks!

musashino205 commented on target/linux/ramips/dts/mt7628an_xiaomi_miwifi-3a.dts:
```diff
@@ -0,0 +1,181 @@
+	wifi@0,0 {
+		compatible = "mediatek,mt76";
+		reg = <0x0000 0 0 0 0>;
+		mediatek,mtd-eeprom = <&factory 0x8000>;
```
ditto

minax007 replied:
Thanks for your fast response!
I have adapted the file accordingly. 
Please check if anything else needs to be changed. 
Thanks!

minax007 commented:
Hi @musashino205 

I see that one of the automated tests failed. 
The failure is not related to the code in this pull request. 
![grafik](https://github.com/user-attachments/assets/ad48b2e5-c707-4731-af21-4321702c29f3)

Do you have any idea how to resolve this?

hauke commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -1183,6 +1183,15 @@ define Device/xiaomi_mi-router-4c
+  DEVICE_VENDOR := Xiaomi
+  DEVICE_MODEL := MiWiFi 3A
+  DEVICE_PACKAGES := kmod-mt76x2
+  SUPPORTED_DEVICES += miwifi-3a
```
`SUPPORTED_DEVICES` should not be needed.

You need it hen you want to do sysupgrade from a different OpenWrt fork which named the device `miwifi-3a` and not `xiaomi_miwifi-3a`.

If this was not added intentionally please remove this line.

minax007 replied:
Hi @hauke 
thanks for the review and the explanation!
I have deleted the line.

Commit:
ramips: Add support for Xiaomi MiWiFi 3A

The Xiaomi MiWiFi 3A wireless router has a similar system architecture as the Xiaomi Mi 4A router, which is already officially supported by OpenWrt.

Product website: https://www.mi.com/miwifi3a

Device specification
--------------------
SoC:      MT7628AN MIPS_24KEc @ 580 MHz 2.4G-bgn 2x2
WiFi:     MT7612EN 5G-an, ac 80 MHz 2T2R
Flash:    16 MB
DRAM:     64 MB
Switch:   MT7628AN (integrated in SoC)
Ethernet: 1 x 10 /100 Mbps
USB:      None
Antennas: 2 x 2,4 GHz and 2 x 5 GHz (all are external and non-detachable)
LEDs:     blue/red/amber
Buttons:  Reset
Serial:   115200,8n1

MAC addresses as verified by OEM firmware:
------------------------------------------
use address source
LAN *:DD factory 0x28
WAN *:DD factory 0x28
2g *:DE factory 0x4
5g *:DF factory 0x8004

OEM firmware uses VLAN's to create the network interface for WAN and LAN.

Bootloader info:
----------------
The stock bootloader uses a "Dual ROM Partition System".
OS1 is a deep copy of OS2.
The bootloader starts OS2 by default.
To force start OS1 it is needed to set "flag_try_sys2_failed=1".

How to install:
---------------
1- Use OpenWRTInvasion to gain Telnet, SSH and FTP access: https://github.com/acecilia/OpenWRTInvasion
[IP: 192.168.31.1 | Username: root | Password: root | FTP-Port: 21]
2- Connect to router using telnet or ssh.
3- Backup all partitions. Use command "dd if=/dev/mtd0 of=/tmp/mtd0". Copy /tmp/mtd0 to computer using ftp.
4- Copy openwrt-ramips-mt76x8-xiaomi_miwifi-3a-squashfs-sysupgrade.bin to /tmp in router using ftp.
5- Enable UART access and change start image to OS1.

nvram set uart_en=1
nvram set flag_last_success=1
nvram set boot_wait=on
nvram set flag_try_sys2_failed=1
nvram commit

6- Erase OS1 & OS2 and install OpenWrt

mtd erase OS1
mtd erase OS2
mtd -r write /tmp/openwrt-ramips-mt76x8-xiaomi_miwifi-3a-squashfs-sysupgrade.bin OS1

Credits:
--------
This PR is based on the work of Zehao Zhang (Github: @ZZH-Finalize) that he had published in the PR: #15698

Signed-off-by: Olgun Demir <olgun.demir@mail.com.tr>
Link: https://github.com/openwrt/openwrt/pull/18427
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Neustradamus commented:
@minax007: Good job for your PR and @hauke for the merging!

It is possible to backport in 24.10?

minax007 commented:
Hi @hauke and @musashino205 
Thank you very much!

Hi @Neustradamus 
do you know if I need to do a new pull request in the 24.10 branch, or if there another way of getting it backported?

minax007 commented:
Hi @Neustradamus 
I have created a pull request so that the router does also get supportet with OpenWrt 24.10:
https://github.com/openwrt/openwrt/pull/18550
